## Dependencies and Data

We will be continuing the same procedures until we used RFE in the last section. Instead of that here we will be using SelectFroModel and for keeping it simple we will use only linear models.

Applying SelectFromModel method working on linear model is after using L1 norm.

In [4]:
%matplotlib notebook
import pandas as pd
import seaborn as sn
import numpy as np
from sklearn.neural_network import MLPClassifier   #Multi-Layer Perceptron Classifier
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix
import matplotlib.pyplot as plt     

dataset=pd.read_csv('heart-disease-data.csv')

print(dataset.head(5))     #prints first 5 values for all columns

array=dataset.values
data=array[:,0:13]
labels=array[:,13]
#model1=MLPClassifier(activation='relu',solver='lbfgs',alpha=1e-5,random_state=1,) #doesnot work with neural networks
model2=LinearSVC(kernel='linear')    
model3=LogisticRegression()

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  


## Splitting Data and Fitting Data

In the Select from model case, we need to first fit the model to some algorithm (preferably SVC or Logistic Regression).

Then select the important features which gives best accuracy when fed to that model.

In [25]:
x_train,x_test,y_train,y_test=train_test_split(data,labels,test_size=0.2,random_state=4)


model2=LinearSVC(C=0.01, penalty="l1", dual=False)
model2.fit(x_train,y_train)
k2=model2.predict(x_test)
print("Accuracy of the SVC model on unseen data is "+str(accuracy_score(k2,y_test)*100)+" %")

model3=LogisticRegression(C=0.01)
model3.fit(x_train,y_train)
k3=model3.predict(x_test)
print("Accuracy of the Logistic regression model on unseen data is "+str(accuracy_score(k3,y_test)*100)+" %")

Accuracy of the SVC model on unseen data is 78.68852459016394 %
Accuracy of the Logistic regression model on unseen data is 81.9672131147541 %


## Feature selection using SelectFromModel

We are using the sklearn in-built model SelectFromModel.

We try this on the models SVC and Logistic regression.

    

In [8]:
from sklearn.feature_selection import SelectFromModel

mainModel1=SelectFromModel(model2, prefit=True)
mainModel2=SelectFromModel(model3, prefit=True)

## Transforming the original data 

In [13]:
print("Original data shape is : ",data.shape)

data_new1=mainModel1.transform(data)
print("New data shape after applying SelectFromModel on SVC : ",data_new1.shape)

data_new2=mainModel2.transform(data)
print("New data shape after applying SelectFromModel on Logistic Regression : ",data_new2.shape)

Original data shape is :  (303, 13)
New data shape after applying SelectFromModel on SVC :  (303, 7)
New data shape after applying SelectFromModel on Logistic Regression :  (303, 5)


As we can see the column value which represents the features has reduced to selecting the best of features

With SVMs and logistic-regression, the parameter C controls the sparsity: the smaller C the fewer features selected. With Lasso, the higher the alpha parameter, the fewer features selected.

## Fitting LinearSVC on the Transformed data

In [51]:
data1=data_new1[:,0:6]
labels1=data_new1[:,6]

x_train1,x_test1,y_train1,y_test1=train_test_split(data,labels,test_size=0.2,random_state=4)

model=LinearSVC()
model.fit(x_train1,y_train1)

k=model.predict(x_test)

print("Accuracy of the model on unseen transformed data : "+str(accuracy_score(k,y_test)*100)+" %")

Accuracy of the model on unseen transformed data : 90.1639344262295 %


As we can see the transformed model has better accuracy than the model trained on normal data trained on LinearSVC which gave 78.68852459016394 % accuracy....Well here i'm just lucky ;p..Normally it is not the case, with feature reduction you guarentee the model won't be subjected to overfitting...This i'll just keep aside to preserve my luck...Please understand..

In [67]:
data2=data_new2[:,0:4]
labels2=data_new1[:,4]

x_train2,x_test2,y_train2,y_test2=train_test_split(data,labels,test_size=0.2,random_state=4)

model=LinearSVC()
model.fit(x_train1,y_train1)

k=model.predict(x_test)

print("Accuracy of the model on unseen transformed data : "+str(accuracy_score(k,y_test)*100)+" %")

Accuracy of the model on unseen transformed data : 91.80327868852459 %


Here too we see a rise in accuracy from 81.9672131147541 % to 91.80327868852459 %..Again here i tried a lot to get the accuracy score above the accuracy score without using feature reduction..This is not always the case,Sometimes you may never reach a better accuracy but one thing for sure is that feature reduction saves the model from overfitting.